src/util/mod.rs

```diff
@@ -99,3 +99,92 @@ pub struct BufferInitDescriptor<'a> {
     /// will panic.
     pub usage: crate::BufferUsage,
 }
+
+/// Utility methods not meant to be in the main API.
+pub trait QueueExt {
+    /// Upload an entire texture and its mipmaps from a source buffer.
+    ///
+    /// Expects all mipmaps to be tightly packed in the data buffer.
+    ///
+    /// If the texture is a 2DArray texture, uploads each layer in order, expecting
+    /// each layer and its mips to be tightly packed.
+    fn write_whole_texture_mips(
+        &self,
+        texture: MipmappedTextureCopyView,
+        data: &[u8],
+        size: crate::Extent3d,
+    );
+}
+
+impl QueueExt for crate::Queue {
+    fn write_whole_texture_mips(
+        &self,
+        texture: MipmappedTextureCopyView<'_>,
+        data: &[u8],
+        size: crate::Extent3d,
+    ) {
+        let format_info = texture.format.describe();
+
+        let (layer_iterations, mip_extent) =
+            if texture.texture_dimension == crate::TextureDimension::D3 {
+                (1, size)
+            } else {
+                (size.depth, crate::Extent3d { depth: 1, ..size })
+            };
+
+        let mut binary_offset = 0;
+        for layer in 0..layer_iterations {
+            for mip in 0..texture.mip_count {
+                let mip_size = mip_extent.at_mip_level(mip).unwrap();
+
+                // When uploading mips of compressed textures and the mip is supposed to be
+                // a size that isn't a multiple of the block size, the mip needs to be uploaded
+                // as it's "physical size" which is the size rounded up to the nearest block size.
+                let mip_physical = mip_size.physical_size(texture.format);
+
+                // All these calculations are performed on the physical size as that's the
+                // data that exists in the buffer.
+                let width_blocks = mip_physical.width / format_info.block_dimensions.0 as u32;
+                let height_blocks = mip_physical.height / format_info.block_dimensions.1 as u32;
+
+                let bytes_per_row = width_blocks * format_info.block_size as u32;
+                let data_size = bytes_per_row * height_blocks;
+
+                let end_offset = binary_offset + data_size as usize;
+
+                self.write_texture(
+                    crate::TextureCopyView {
+                        texture: &texture.texture,
+                        mip_level: mip as u32,
+                        origin: crate::Origin3d {
+                            x: 0,
+                            y: 0,
+                            z: layer,
+                        },
+                    },
+                    &data[binary_offset..end_offset],
+                    crate::TextureDataLayout {
+                        offset: 0,
+                        bytes_per_row,
+                        rows_per_image: 0,
```

Comment:
this, in particular, is demonstrating the benefit of iterating over the mipmaps in the outer loop. This way, you'd not need to iterate the layers at all, and just upload all layers at once.